### PR TITLE
HHH-12780 - Re-allow native SQL queries that take a given resultClass to map the result set to the required type

### PR DESCRIPTION
**Pull request by vladmihalcea**
https://hibernate.atlassian.net/browse/HHH-12780

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java [hunk 1]
```diff
@@ -932,12 +934,36 @@ public NativeQueryImplementor createNativeQuery(String sqlString, Class resultCl
...
 			query.addEntity( "alias1", resultClass.getName(), LockMode.READ );
 		}
+		else if ( !isObjectArray ) {
+			query.setResultTransformer( new BasicTransformerAdapter() {
+				@Override
+				public Object transformTuple(Object[] tuple, String[] aliases) {
+					return resultClass.cast( tuple[0] );
+				}
+			} );
+		}
```
Comment (gsmet) [inline hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java]: What in the `else` case? Is it legit to not do anything?

Comment (vladmihalcea) [reply hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java]: In that case, the result is processed as `Object[]` and we don't have to do any transformation.

**Comment by gsmet** [comment]
@vladmihalcea so about this one, I also reverted the tests you added so you would have to add them back.

File: hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java [hunk 2]
```diff
@@ -932,12 +934,36 @@ public NativeQueryImplementor createNativeQuery(String sqlString, Class resultCl
...
+				public Object transformTuple(Object[] tuple, String[] aliases) {
+					Object[] result = (Object[]) Array.newInstance( elementClass, tuple.length );
+					for ( int i = 0; i < tuple.length; i++ ) {
+						result[i] = elementClass.cast( tuple[i] );
+					}
+					return result;
+				}
+			} );
+		}
+		else if ( this.getFactory().getMetamodel().entityPersisters().get( resultClass.getName() ) != null ) {
```
Comment (gsmet) [inline hibernate-core/src/main/java/org/hibernate/internal/AbstractSharedSessionContract.java]: Maybe I'm wrong but I think the key of the map returned by the `entityPersisters()` is not always the class name but can also be the entity name if defined (at least if the mapping is done via XML, it seems that there is a possibility the class name is not used).

See https://github.com/hibernate/hibernate-orm/blob/master/hibernate-core/src/main/java/org/hibernate/boot/model/source/internal/hbm/AbstractEntitySourceImpl.java#L114

That being said, it looks like `MetamodelImpl#entityPersister(Class entityClass)` is doing the same mistake so I'm not sure about it.

**Comment by gsmet** [comment]
@vladmihalcea still requires the tests to be reintroduced: https://github.com/hibernate/hibernate-orm/commit/9fac6747ef22acff9c7350efca33286452a2c967#diff-8366ba4b8b14c099a596fe79b1ab7849

**Comment by vladmihalcea** [comment]
This will be implemented in 6.0.

**Pull request closed without merging**